File: frontend/src/components/form/schemas/interpreter.schema.ts
import * as yup from 'yup'
import { BookingType } from 'components/form/schemas/booking.schema'
import { Language } from 'constants/interfaces'

const Initial = {
  firstName: '',
  lastName: '',
  languages: [{
    languageName: '',
    level: '',
    commentOnLevel: ''
  }],
  address: '',
  city: '',
  province: 'BC',
  postal: '',
  homePhone: '',
  businessPhone: '',
  phone: '',
  email: '',
  supplier: '',
  gst: '',
  comments: '',
  contractExtension: false,
  contractTermination: false,
}

const Schema = yup.object({
  firstName: yup.string().required('First name is required'),
  lastName: yup.string().required('Last name is required'),
  languages: yup.array().of(yup.object({
    languageName: yup.string().required('Please provide a language name'),
    level: yup.string().matches(/(1|2|3|4)/, 'Level must be a number from 1-4'),
    comment: yup.string(),
  })),
  address: yup.string(),
  city: yup.string(),
  province: yup.string(),
  postal: yup.string(),
  homePhone: yup.string(),
  businessPhone: yup.string(),
  phone: yup.string(),
  email: yup.string().email('Please provide a valid email'),
  supplier: yup.string(),
  gst: yup.string(),
  comments: yup.string(),
  contractExtension: yup.boolean(),
  contractTermination: yup.boolean(),
})

export type Interpreter = {
  id: number;
  firstName: string;
  lastName: string;
  languages: Language[];
  bookings: BookingType[];
  address?: string;
  city?: string;
  province?: string;
  postal?: string;
  homePhone?: string;
  businessPhone?: string;
  phone?: string;
  email?: string;
  supplier?: string;
  gst?: string;
  comments?: string;
  contractExtension?: boolean;
  contractTermination?: boolean;
  createdAt?: Date;
  updatedAt?: Date;
}

export { Schema, Initial }

File: api/api/schemas/interpreter_schema.py
from datetime import datetime
from typing import Optional, List
from pydantic import BaseModel, Field

from api.schemas.language_schema import InterpreterLanguageSchema
from api.schemas.custom_type import JsonBase
from api.schemas.location_schema import CourtDistanceSchema


#__________________________________________
#__________________________________________
class InterpreterBase(BaseModel):

    last_name: Optional[str] = Field(alias="lastName")
    first_name: Optional[str] = Field(alias="firstName")

    address: Optional[str]
    city: Optional[str]
    province: Optional[str] = "BC"
    postal_code: Optional[str] = Field(alias="postal")
    
    home_phone: Optional[str] = Field(alias="homePhone")
    business_phone: Optional[str] = Field(alias="businessPhone")
    cell_phone: Optional[str] = Field(alias="phone")
    fax: Optional[str] = None    
    email: Optional[str] =''

    supplier_no: Optional[str] = Field(alias="supplier")
    gst_no: Optional[str] = Field(alias="gst")
    site_code: Optional[str] = Field(default=None, alias="siteCode")
    
    contract_valid: Optional[bool] = Field(None, alias="contractExtension")
      
    languages: Optional[List[InterpreterLanguageSchema]] = []

    comments: Optional[str] = None
    
    class Config():
        orm_mode = True
        allow_population_by_field_name = True
#__________________________________________
#__________________________________________

class InterpreterCreateModifyRequestSchema(InterpreterBase):
    crc_check_date: Optional[datetime] = Field(alias="criminalRecordCheckDate")
    completed_training: Optional[bool] = False
    admin_comment: Optional[str] = Field("", alias="adminComments")
    crc_comment: Optional[str] = Field(None, alias="criminalRecordCheckComment")
    contract_comment: Optional[str] = None



class InterpreterGetAdminResponseSchema(InterpreterCreateModifyRequestSchema):
    id: int     
    events: Optional[List] = []
    booking: Optional[List] = []
    created_at: Optional[datetime]



# payload_in_page_directory = criminalRecordCheck":"2021-12-25T03:39:00.000Z"} "Time is back utc. selected 2021-12-24T19:39Vancouver"




class InterpreterBookingResponseSchema(BaseModel):
    id: Optional[int] 
    last_name: Optional[str] = Field(alias="lastName")
    first_name: Optional[str] = Field(alias="firstName")
    cell_phone: Optional[str] = Field(alias="phone")
    email: Optional[str]
    languages: Optional[List[InterpreterLanguageSchema]] = []
    language_history: Optional[JsonBase] = Field(alias="languageHistory")
    courts: Optional[List[CourtDistanceSchema]]
    address: Optional[str]
    city: Optional[str]
    province: Optional[str] = "BC"
    postal_code: Optional[str] = Field(alias="postal")

    class Config():
        orm_mode = True
        allow_population_by_field_name = True


class InterpreterADMBookingResponseSchema(InterpreterBookingResponseSchema):
    supplier_no: Optional[str] = Field(alias="supplier")
    gst_no: Optional[str] = Field(alias="gst")
    site_code: Optional[str] = Field(alias="siteCode")
    address_longitude : Optional[float] = Field(alias="addressLongitude")
    address_latitude : Optional[float] = Field(alias="addressLatitude")


class InterpreterGeoStatusSchema(BaseModel):
    id: int
    update_started = False
    last_name: Optional[str] = Field(alias="lastName")
    first_name: Optional[str] = Field(alias="firstName")

    address: Optional[str]
    city: Optional[str]
    province: Optional[str] = "BC"
    postal_code: Optional[str] = Field(alias="postal")

    updated_at: Optional[datetime]
    contract_valid: Optional[bool] = Field(None, alias="contractExtension")
    geo_service: Optional[str]

    class Config():
        orm_mode = True
        allow_population_by_field_name = True

class InterpreterBookingResponseShortSchema(BaseModel):
    id: Optional[int] 
    last_name: Optional[str] = Field(alias="lastName")
    first_name: Optional[str] = Field(alias="firstName")
    class Config():
        orm_mode = True
        allow_population_by_field_name = True